torch/ao/quantization/_pt2e/prepare.py

```diff
@@ -0,0 +1,115 @@
+    nodes_before_observation = list(model.graph.nodes)
+
+    for node in nodes_before_observation:
+
+        if node.op == "placeholder":
+            pass
+        elif node.op in ("call_module", "call_method", "call_function"):
+            this_node_dtype_info = node.meta["target_dtype_info"]
+            if "val" in node.meta:
+                output_is_a_tensor = (
+                    this_node_dtype_info is not None and
+                    isinstance(node.meta["val"], FakeTensor)
+                )
+            else:
+                output_is_a_tensor = this_node_dtype_info is not None
+
+            skip_inserting_observers = (
+                not output_is_a_tensor
+            )
+
+            if skip_inserting_observers:
+                continue
+
+            _maybe_insert_input_observers_for_node(
+                node,
+                None, # qconfig
+                model,
+                named_modules,
+                model.graph,
+                None, # qhandler
+                PrepareCustomConfig(),
+                None, # backend_config
+            )
+
+            named_modules = dict(model.named_modules(remove_duplicate=False))
+            # this returns the new observer node if it was needed
+            maybe_output_obs_node = _maybe_insert_output_observer_for_node(node, model, named_modules, model.graph)
+
+            if maybe_output_obs_node is None:
+                continue
+            # Update users of original node to use the output observer
+            # instead. For example, change
+            #
+            #           next_node
+            #          /
+            #   cur_node -> obs
+            #
+            # to
+            #
+            #                 next_node
+            #                 /
+            #   cur_node -> obs
+            #
+            # We need to save orig users before updating uses because
+            # the list of users will change as we update uses
+            orig_users = list(node.users.keys())
+            for user_node in orig_users:
+                if user_node is maybe_output_obs_node:
+                    continue
+                user_node.replace_input_with(node, maybe_output_obs_node)
+            _is_observer_in_same_graph_ = _is_observer_in_same_graph(node, named_modules)
+
+            # for general tensor value ops, we modify the graph
+            # to make all inputs and outputs use the first input's
+            # observer
+            if (is_general_tensor_value_op and _is_observer_in_same_graph_) or \
+                    _is_reuse_input_qconfig_:
+                if not _maybe_make_input_output_share_observers(node, model, named_modules):
+                    _remove_output_observer(node, model, named_modules)
+
+        elif node.op == "output":
+            _maybe_insert_observers_before_graph_output(
+                node, model, named_modules, model.graph)
+
+    model = GraphModule(model, model.graph)
+
+    _save_state(
```

Comment:
Why save_state?

Reply:
convert code is expecting this, a lot of things are actually not needed for pt2e flow, but we can clean up later I think

Reply:
fx graph mode quantization has more features than pt2e right now, e.g. qat support, standalone_module, custom_module, it is also working with modules, a lot of states are for supporting these features